### PR TITLE
[JCLOUDS-1255] Initial API structure and get account API implementation

### PR DESCRIPTION
**Pull request by jawnclarke**
- Adding structure for Dimension Data CloudControl API JClouds implementation
- Adding single API for user account retrieval, with mock and live tests
- Adding HTTP request filter that adds organization value to API request paths

### CODE REVIEW

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import org.jclouds.dimensiondata.cloudcontrol.DimensionDataCloudControlApiMetadata;
+import org.jclouds.sshj.config.SshjSshClientModule;
+import org.testng.annotations.Test;
+
+import java.util.Properties;
+
+@Test(groups = "live")
+public class BaseDimensionDataCloudControlApiLiveTest extends BaseApiLiveTest<DimensionDataCloudControlApi> {
+
+   protected String ORG_ID = System
```
Comment (jawnclarke) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java]: Thanks, Trevor! I've removed it.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      this(new Builder());
+   }
+
+   protected DimensionDataCloudControlApiMetadata(Builder builder) {
+      super(builder);
+   }
+
+   public static Properties defaultProperties() {
+      Properties properties = BaseHttpApiMetadata.defaultProperties();
+      properties.setProperty(Constants.PROPERTY_CONNECTION_TIMEOUT, "1200000"); // 15 minutes
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java]: Can we avoid setting this? The connection timeout configures the time to get a connection opened, not the time required to read all the data. It looks excessive. Let's better remove this default and let users configure it as needed.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      protected Builder() {
+         id("dimensiondata-cloudcontrol").name("DimensionData CloudControl API").identityName("user name")
+               .credentialName("user password")
+               .documentation(URI.create("http://www.dimensiondata.com/en-US/Solutions/Cloud"))
+               .defaultEndpoint("https://api-REGION.dimensiondata.com/").version("2.4")
+               .defaultProperties(DimensionDataCloudControlApiMetadata.defaultProperties())
+               .view(typeToken(ComputeServiceContext.class)).defaultModules(
+               ImmutableSet.<Class<? extends Module>>builder().add(DimensionDataCloudControlHttpApiModule.class)
+                     .add(DimensionDataCloudControlParserModule.class)
+                     .add(DimensionDataComputeServiceContextModule.class).build());
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java]: For now let's remove the ComputeServiceAdapter module, the implementation and all other transformation functions. They are not implemented and return `null` and placeholders, and provide no value and just add complexity to the code and the PR. Let's remove all that stuff and start with a clean API to set a good foundation for the provider.
That will help and allow further PRs to be easier and faster.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   @Override
+   protected void bindRetryHandlers() {
+      //TODO Implement Method
+   }
+
+   @Override
+   protected void installLocations() {
+      install(new LocationModule());
+      bind(ImplicitLocationSupplier.class).to(OnlyLocationOrFirstZone.class).in(Scopes.SINGLETON);
+   }
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java]: Don't override this yet. Let's remove it for now and we'll add it back (if it turns needed) when implementing the compute service adapter.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+   @Override
+   protected void bindErrorHandlers() {
+      //TODO Implement Method
+   }
+
+   @Override
+   protected void bindRetryHandlers() {
+      //TODO Implement Method
+   }
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java]: You can remove this, as jclouds provides sane default retry handlers.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import org.jclouds.rest.ConfiguresHttpApi;
+import org.jclouds.rest.config.HttpApiModule;
+
+@ConfiguresHttpApi
+public class DimensionDataCloudControlHttpApiModule extends HttpApiModule<DimensionDataCloudControlApi> {
+
+   @Override
+   protected void bindErrorHandlers() {
+      //TODO Implement Method
+   }
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java]: Remove this method if you're not defining any custom error handler.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlParserModule.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import org.jclouds.Constants;
+import org.jclouds.json.config.GsonModule;
+
+import java.util.Collections;
+
+public class DimensionDataCloudControlParserModule extends AbstractModule {
+
+   @Override
+   protected void configure() {
+      Names.bindProperties(binder(), Collections.singletonMap(Constants.PROPERTY_PRETTY_PRINT_PAYLOADS, "true"));
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlParserModule.java]: There is no need to bind this. It is already bound to the context and users will be able to configure this property when creating it. Let's remove this line.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java [hunk 8]
```diff
@@ -44,36 +44,36 @@
...
+   public abstract AccountOrganization organization();
+
+   @SerializedNames({ "userName", "fullName", "firstName", "lastName", "emailAddress", "roles", "phone", "department",
+         "customDefined1", "customDefined2", "organization", "state" })
+   public static Account create(String userName, String fullName, String firstName, String lastName,
+         String emailAddress, List<RoleType> roles, AccountPhoneNumber phone, String department, String customDefined1,
+         String customDefined2, AccountOrganization organization, String state) {
+      return builder().userName(userName).fullName(fullName).firstName(firstName).lastName(lastName)
+            .emailAddress(emailAddress)
+            .roles(roles == null ? ImmutableList.<RoleType>of() : ImmutableList.copyOf(roles)).phone(phone)
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java]: Just pass the list here as it comes from the parameter, and move the immutability logic to the builder following [this pattern](https://github.com/jclouds/jclouds-labs/blob/master/dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Backup.java#L53-L62).

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java [hunk 9]
```diff
@@ -44,36 +44,36 @@
...
 
-   public abstract String orgId();
+   @Nullable
+   public abstract String state();
 
    @Nullable
-   public abstract Roles roles();
+   public abstract List<RoleType> roles();
 
-   Account() {
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java]: Why has this been removed? Providing a non-visible default constructor is a best-practice to enforce using the builders. Let's add the constructors back.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java [hunk 10]
```diff
@@ -44,36 +44,36 @@
...
+   @Nullable
+   public abstract AccountPhoneNumber phone();
+
+   public abstract AccountOrganization organization();
+
+   @SerializedNames({ "userName", "fullName", "firstName", "lastName", "emailAddress", "roles", "phone", "department",
+         "customDefined1", "customDefined2", "organization", "state" })
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/domain/Account.java]: jclouds makes some assumptions when serializing and deserializing. The order of the parameters in this annotation **must be the same** than the order in which the variables are declared in the class. Please, reorder accordingly.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilter.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+ * Handles both oec and caas based URLs.
+ */
+@Singleton
+public class OrganisationIdFilter implements HttpRequestFilter {
+
+   private static final int ORGANIZATION_ID_INDEX = 3;
+   private final Supplier<String> organisationIdSupplier;
+
+   @Inject
+   public OrganisationIdFilter(@Memoized Supplier<String> organisationIdSupplier) {
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilter.java]: Remove the `public` modifier. It is a best-practice in Guice to make all injection constructors visible just to the default scope.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/config/DimensionDataComputeServiceContextModule.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+   }
+
+   @Provides
+   @Singleton
+   @Memoized
+   public final Supplier<String> getOrganisationIdForAccount(AtomicReference<AuthorizationException> authException,
+         @Named(PROPERTY_SESSION_INTERVAL) long seconds, DimensionDataCloudControlApi api) {
+      return MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier
+            .create(authException, new OrganisationIdForAccount(api), seconds, TimeUnit.SECONDS);
+   }
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/config/DimensionDataComputeServiceContextModule.java]: Better move this supplier to the HttpApi module, since it is generic for the provider APIs, not specifically about the compute abstraction.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/suppliers/OrganisationIdForAccount.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+@Singleton
+public class OrganisationIdForAccount implements Supplier<String> {
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final AccountApi api;
+
+   @Inject
+   public OrganisationIdForAccount(DimensionDataCloudControlApi api) {
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/suppliers/OrganisationIdForAccount.java]: Remove the `public` modifier.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/suppliers/OrganisationIdForAccount.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+import javax.annotation.Resource;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@Singleton
+public class OrganisationIdForAccount implements Supplier<String> {
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/suppliers/OrganisationIdForAccount.java]: better remove if unused.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilter.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+      this.organisationIdSupplier = organisationIdSupplier;
+   }
+
+   @Override
+   public HttpRequest filter(HttpRequest request) throws HttpException {
+      return request.toBuilder().replacePath(injectOrganisationId(request.getEndpoint().getPath())).build();
+   }
+
+   @VisibleForTesting
+   public String injectOrganisationId(String path) {
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilter.java]: Don't make it public if it is just visible for testing. Remove the public modifier and leave it in default scope.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+      assertThat(account.roles().contains(Account.RoleType.create("storage")));
+      assertThat(account.roles().contains(Account.RoleType.create("primary administrator")));
+      assertEquals("1", account.phone().countryCode());
+      assertEquals("1234567", account.phone().number());
+      assertEquals("6ac1e746-b1ea-4da5-a24e-caf1a978789d", account.organization().id());
+      assertEquals("Eng Customer of DevLab1-homed online sign-up Vendor", account.organization().name());
+      assertEquals("DevLab1Geo1", account.organization().homeGeoName());
+      assertEquals("apidevlab1.opsourcecloud.net", account.organization().homeGeoApiHost());
+      assertEquals("DEVLAB1GEO1", account.organization().homeGeoId());
+      assertEquals("NORMAL", account.state());
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java]: Values seem to be of a particular user. Change them for `notNull` asserts or something that can be run by anyone.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.jclouds.dimensiondata.cloudcontrol.domain.Account;
+import org.jclouds.dimensiondata.cloudcontrol.internal.BaseDimensionDataCloudControlApiLiveTest;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.AssertJUnit.assertEquals;
+
+@Test(groups = "live", testName = "AccountApiLiveTest", singleThreaded = true)
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java]: Looks like it does not need to be single-threaded.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+import static org.testng.AssertJUnit.assertEquals;
+
+@Test(groups = "unit", testName = "AccountApiMockTest", singleThreaded = true)
+public class AccountApiMockTest extends BaseDimensionDataCloudControlMockTest {
+
+   @BeforeMethod
+   @Override
+   public void start() throws IOException {
+      super.start();
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java]: No need to override this.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+      assertEquals("1234567", account.phone().number());
+      assertEquals("Marketing", account.department());
+      assertEquals("custom defined 1", account.customDefined1());
+      assertEquals("custom defined 2", account.customDefined2());
+      assertEquals("6ac1e746-b1ea-4da5-a24e-caf1a978789d", account.organization().id());
+      assertEquals("Eng Customer of DevLab1-homed online sign-up Vendor", account.organization().name());
+      assertEquals("DevLab1Geo1", account.organization().homeGeoName());
+      assertEquals("apidevlab1.opsourcecloud.net", account.organization().homeGeoApiHost());
+      assertEquals("DEVLAB1GEO1", account.organization().homeGeoId());
+      assertEquals("NORMAL", account.state());
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java]: There is no need to verify the response deserialization in mock tests, so you can remove all these asserts. We just want to make sure the generated requests are the expected ones.
That method uses basic authentication (via its filter), so let's modify the `assertSent` method in the base mock test class to verify that the authentication header is sent too.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+      assertSent(server, "GET", "/caas/2.4/user/myUser");
+   }
+
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiMockTest.java]: Add another test method that simulates a 404 response, to verify that the method returns `null`.
Since you are defining an explicit `@Fallback` in the api method, we'd better verify it behaves as expected on the affected responses.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+import org.jclouds.dimensiondata.cloudcontrol.internal.BaseDimensionDataCloudControlMockTest;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+import static org.testng.Assert.assertEquals;
+
+@Test(groups = "unit", testName = "OrganisationIdFilterTest", singleThreaded = true)
+public class OrganisationIdFilterTest extends BaseDimensionDataCloudControlMockTest {
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java]: There is no need for this test to be single-threaded nor to extend the base mock test class. You don't need to actually start a mock web server to test this class.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+   @BeforeMethod
+   @Override
+   public void start() throws IOException {
+      super.start();
+      server.enqueue(xmlResponse("/account.json"));
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java]: Also remove this.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+   @Test
+   public void testCaasUrl() {
+      Supplier<String> orgIdSupplier = new OrganisationIdForAccount(api);
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java]: You can just return a mock supplier for this test: `Suppliers.ofInstance("6ac1e746-b1ea-4da5-a24e-caf1a978789d/")`.
This way you don't need the overhead of starting a mock server, creating the context, etc.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+   }
+
+   @Test
+   public void testDontAddIfAlreadyPresent() {
+      Supplier<String> orgIdSupplier = new OrganisationIdForAccount(api);
+      String expectedPath = "/caas/2.4/6ac1e746-b1ea-4da5-a24e-caf1a978789d/server/0896551e-4fe3-4450-a627-ad5548e7e83a";
+      String updatedPath = new OrganisationIdFilter(orgIdSupplier).injectOrganisationId(
+            "/caas/2.4/6ac1e746-b1ea-4da5-a24e-caf1a978789d/server/0896551e-4fe3-4450-a627-ad5548e7e83a");
+      assertEquals(updatedPath, expectedPath);
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/filters/OrganisationIdFilterTest.java]: Also add a test to verify the behavior when the path segments are < 3 ?

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+   @Override
+   protected ApiMetadata createApiMetadata() {
+      return new DimensionDataCloudControlApiMetadata();
+
+   }
+
+   @Override
+   protected Iterable<Module> setupModules() {
+      return ImmutableSet.<Module>of(getLoggingModule(), new SshjSshClientModule());
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java]: The SSH module is only needed for the compute service tests, so you could remove this override method.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+   }
+
+   @Override
+   protected Properties setupProperties() {
+      Properties overrides = super.setupProperties();
+      overrides.setProperty("jclouds.ssh.retry-auth", "true");
+      return overrides;
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlApiLiveTest.java]: SSH config is only needed for compute tests, so you should be able to remove this method too.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlMockTest.java [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+      return server.getUrl(path).toString();
+   }
+
+   protected MockResponse jsonResponse(String resource) {
+      return new MockResponse().addHeader("Content-Type", "application/json").setBody(stringFromResource(resource));
+   }
+
+   protected MockResponse xmlResponse(String resource) {
+      return new MockResponse().addHeader("Content-Type", "application/xml").setBody(stringFromResource(resource));
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlMockTest.java]: Remove this? Looks like we're just consuming a json api.

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlMockTest.java [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+
+   protected MockResponse responseUnexpectedError() {
+      return new MockResponse().setResponseCode(400).setStatus("HTTP/1.1 400 Bad Request")
+            .setBody("content: [{\"operation\":\"OPERATION\",\"responseCode\":\"UNEXPECTED_ERROR\"}]");
+   }
+
+   protected MockResponse responseResourceNotFound() {
+      return new MockResponse().setResponseCode(400).setStatus("HTTP/1.1 400 Bad Request")
+            .setBody("content: [{\"operation\":\"OPERATION\",\"responseCode\":\"RESOURCE_NOT_FOUND\"}]");
+   }
```
Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/internal/BaseDimensionDataCloudControlMockTest.java]: Hmmm, upon a 400 response, is this what the DimensionData API returns? Do clients need to actually read the response body to really know the cause of the failure?

In any case, it looks like we need a custom error handler to propagate the right exception upon failed requests. jclouds by default throws `HttpResponseExceptions`, but there are other exceptions that better represent the different response codes:

* IllegalArgumentException - for 400 Bad Request or 409 Conflict.
* ResourceNotFoundException - for 404 Not Found.
* AuthorizationException - for 403 Unauthorized.

Those can be handled by creating a `DimensionDataCloudControlErrorHandler` (you can take [the Packet one](https://github.com/jclouds/jclouds-labs/blob/master/packet/src/main/java/org/jclouds/packet/handlers/PacketErrorHandler.java) as an example) and declaring it in the http api module (as in [here](https://github.com/jclouds/jclouds-labs/blob/master/packet/src/main/java/org/jclouds/packet/config/PacketHttpApiModule.java#L48-L53)).

**Comment by trevorflanagan** [comment]
@nacx thanks for taking the time to look at this. These comments are really helpful for us. @jawnclarke and I will work through the changes you requested.

**Comment by nacx** [comment]
Great! I know it's quite difficult to get started with the jclouds code, so please, don't hesitate to ask anything you need. I'll be happy to help and provide all the details you need so you can understand how everything works. We are also in the #jclouds IRC channel at Freenode, so feel free to join there if you need real time feedback! :)

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java (hunk 3 above)
Comment (trevorflanagan) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java]: @nacx Before making too many changes I want to make sure I am doing the right thing. Your expectation on this commnet is that `DimensionDataComputeServiceContextModule` , `DimensionDataComputeServiceDependenciesModule` `DimensionDataCloudControlComputeServiceAdapter` are removed? As a result of removing those I also remove the unused transformation functions classes.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java (hunk 5 above)
Comment (trevorflanagan) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java]: @nacx I intend on now retaining this method with the binding for the `DimensionDataCloudControlErrorHandler`

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java (hunk 6 above)
Comment (trevorflanagan) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java]: @nacx I intend on now retaining this method with the binding for the DimensionDataCloudControlErrorHandler

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java (hunk 3 above)
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/DimensionDataCloudControlApiMetadata.java]: That's right.That's right. Let's remove the entire "compute" package (moving the org id supplier elsewhere if it is needed).

File: dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java (hunk 16 above)
Comment (jawnclarke) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java]: Good to know this approach going forward. Thanks @nacx .

Comment (nacx) [inline dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/features/AccountApiLiveTest.java]: We have 2 kinds ot tests in jclouds: mock and live tests.

Mock tests verify that we generate the requests we are expected to generate. One could think about them as "jclouds generates exactly what the API docs say".
Live tests are integration tests that perform real requests to the target cloud (and may cost money!). The idea is that anyone with an account should be able to run them to have the real status of the APIs, configure them in the CI to keep track of the changes on the provider API, etc.

File: dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlParserModule.java [hunk 29]
```diff
@@ -17,18 +17,11 @@
...
-import org.jclouds.json.config.GsonModule;
-
-import java.util.Collections;
 
 public class DimensionDataCloudControlParserModule extends AbstractModule {
 
    @Override
    protected void configure() {
-      Names.bindProperties(binder(), Collections.singletonMap(Constants.PROPERTY_PRETTY_PRINT_PAYLOADS, "true"));
-      bind(GsonModule.DateAdapter.class).to(GsonModule.Iso8601DateAdapter.class);
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlParserModule.java]: We'll probably need to bind something like this adapter to be able to deserialize json strings to Date objects. Let's see which format the target API uses.

**Comment by nacx** [comment]
Thanks @jawnclarke!
Mind squashing the commits into a single one so I can cleanly merge this?

**Comment by nacx** [comment]
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/6e5082f7) and [2.0.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/884327fa). Thanks!
